### ISSUE TITLE
"Duplicate FOREIGN KEY constraint name" error when using default value `NULL` in `db_schema.xml` on foreign key column

### ISSUE THREAD

**Issue by Fgruntjes** [magento/magento2#23965]
### Preconditions (*)
1. Magento 2.3.2

### Steps to reproduce (*)
1. Create a module with a `db_schema.xml` containing a nullable foreign key column:

```xml
<?xml version="1.0"?>
<schema xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Setup/Declaration/Schema/etc/schema.xsd">
    <table name="some_table" resource="default" engine="innodb">
        <column xsi:type="smallint" name="store_id" padding="5" unsigned="true"  nullable="true"/>

        <constraint xsi:type="foreign" referenceId="STORE_ID" table="some_table"
                    column="store_id" referenceTable="store" referenceColumn="store_id" onDelete="CASCADE"/>
    </table>
</schema>
```

2. Run the `setup:upgrade` command twice.

### Expected result (*)
1. First run of `setup:upgrade` the table is created
2. Second run should do nothing

### Actual result (*)
1. First run of `setup:upgrade` the table is created
2. Second run results in MySQL error:
```
SQLSTATE[HY000]: General error: 1826 Duplicate FOREIGN KEY constraint name 'magento_test/SOME_TABLE_STORE_ID_STORE_STORE_ID', query was: ALTER TABLE `some_table` MODIFY COLUMN `store_id` smallint(5) UNSIGNED NULL, ADD CONSTRAINT `SOME_TABLE_STORE_ID_STORE_STORE_ID` FOREIGN KEY (`store_id`) REFERENCES `store` (`store_id`)  ON DELETE CASCADE
```

**Comment by Fgruntjes** [comment]
@magento give me 2.3-develop instance - upcoming 2.3.x release

**Comment by magento-engcom-team** [comment]
Hi @Fgruntjes. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @Fgruntjes, here is your [Magento instance](https://i-23965-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23965-2-3-develop.instances.magento-community.engineering/admin](https://i-23965-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by engcom-Charlie** [comment]
Hello @Fgruntjes 
I can't reproduce this issue on fresh Magento 2.3-develop.
**Manual testing scenario:**
1.Create a module with a db_schema.xml containing a nullable foreign key column:
```xml
<?xml version="1.0"?>
<schema xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Setup/Declaration/Schema/etc/schema.xsd">
    <table name="some_table" resource="default" engine="innodb">
        <column xsi:type="smallint" name="store_id" padding="5" unsigned="true"  nullable="true"/>

        <constraint xsi:type="foreign" referenceId="STORE_ID" table="some_table"
                    column="store_id" referenceTable="store" referenceColumn="store_id" onDelete="CASCADE"/>
    </table>
</schema>
```
2.Run setup:upgrade
3.Check a created table in mysql(have to be exist)
4.Run setup:upgrade again
**Actual result:**
Upgrade finished successfully

So i have to close this issue.
Thank for your report!

**Comment by zapotocnylubos** [comment]
I have got **exactly** same error.
magento version: 2.3.2
db: mariadb:10.2.21

**Comment by engcom-Charlie** [comment]
Hello @zapotocnylubos 
I am still can't reproduce it on Magento 2.3.2
**GIF:**
![Peek 2019-09-27 09-39](https://user-images.githubusercontent.com/51680850/65747913-5779ca00-e10b-11e9-8c6b-8a84ac3246ea.gif)

**Comment by andrewbess** [comment]
Hi @engcom-Charlie 
I have reproduced this problem.
Perhaps, the problem is because the name of an index in the database is different from referenceId when you get create declaration schema and db_schema_whitelist.json manually.
For example:
I have table `mm_accounting_invoice` that contains the next declaration:
`<constraint referenceId="ACCOUNTING_INVOICE_STORE_ID_STORE_STORE_ID" xsi:type="foreign" referenceColumn="store_id" onDelete="SET NULL"/>`
So, I created db_schema_whitelist.json with the next contain:

I try to do setup:upgrade and I am getting an error
https://monosnap.com/file/0YheVV9F5A4AS85w2C1s0iN5piyCDb
There are different indexes in DB and db_schema.xml.
I changed it in db_schema_whitelist.json following to indexes named in DB.
It works.
Please take a look at my video:
https://monosnap.com/file/FC8oLV7WOnYgA3bYidnb1OQIj0igog

P.S. @Fgruntjes @zapotocnylubos please use
`bin/magento setup:db-declaration:generate-whitelist --module-name=Vendor_Module`

**Comment by andrewbess** [comment]
@engcom-Charlie 
I have reproduced this problem.
I don't know yet this is bug or no.
So, the problem is possible if file "db_schema_whitelist.json" does not exist or incorrect.
For example:
1. I have the table `mm_accounting_invoice` in DB;
2. This has the following contain:
`<constraint referenceId="ACCOUNTING_INVOICE_STORE_ID_STORE_STORE_ID" xsi:type="foreign" table="mm_accounting_invoice" column="store_id" referenceTable="store" referenceColumn="store_id" onDelete="SET NULL"/>`. Please take note on referenceId;
3. I created "db_schema_whitelist.json" manually with following contain:
`"constraint": {
            "PRIMARY": true,
            "ACCOUNTING_INVOICE_STORE_ID_STORE_STORE_ID": true
        }`
So, during to first run declaration schema creates new table `mm_accounting_invoice`, but when I restart the `bin/magento setup:upgrade` I get an error:
https://monosnap.com/file/FuFnCiWhgHj0ZaNlcGRxRfhWx1bfGK.
Please take a look at my video [here.](https://monosnap.com/file/1bSHelxiXx6N5bJimXWx9b33tLlpje)
This problem is possible due to different names of referenceId and index in the DB. It can be fixed using `bin/magento setup:db-declaration:generate-whitelist --module-name=Vendor_Module`.
So, there will be created correct file "db_schema_whitelist.json" with correct index name (`MM_ACCOUNTING_INVOICE_STORE_ID_STORE_STORE_ID` in my case).
Why don't we use `referenceId` as an index name?

P.S. @Fgruntjes @zapotocnylubos zapotocnylubos You should use `bin/magento setup:db-declaration:generate-whitelist --module-name=Vendor_Module` every time after a change "db_schema.xml".

**Comment by engcom-Charlie** [comment]
Hello @andrewbess 
I still can't reproduce this issue on a fresh Magento 2.3-develop.
Yes, we have to use comman `bin/magento setup:db-declaration:generate-whitelist` after updating db-shema.
So this issue should be closed.
Thank you for contribution and collaboration!

**Comment by toadster122** [comment]
Hello @engcom-Charlie,

I encountered this problem on one of my websites and did debugging to try and discover the underlying issue. In my case, it was caused by editing the column backing the foreign key reference. The edit of that backing column meant that the function turnOffForeignKeys in TableDiff.php attempted to remove and re-add the foreign key, but because that column wasn't in the whitelist, it couldn't do so. 

It seems to me that intended functionality would be to allow the removal and recreation of the foreign key, but it's possible I'm overlooking something about that design.
